tooling/image-sync/internal/sync.go

```diff
@@ -0,0 +1,196 @@
+	policyctx, err := signature.NewPolicyContext(&signature.Policy{
+		Default: signature.PolicyRequirements{
+			signature.NewPRInsecureAcceptAnything(),
+		},
+	})
+	if err != nil {
+		return err
+	}
+
+	src, err := docker.ParseReference("//" + srcreference)
+	if err != nil {
+		return err
+	}
+
+	dst, err := docker.ParseReference("//" + dstreference)
+	if err != nil {
+		return err
+	}
+
+	_, err = copy.Image(ctx, policyctx, dst, src, &copy.Options{
+		SourceCtx: &types.SystemContext{
+			DockerAuthConfig: srcauth,
+		},
+		DestinationCtx: &types.SystemContext{
+			DockerAuthConfig: dstauth,
+		},
+	})
+
+	return err
+}
+
+func readQuaySecret(filename string) (*QuaySecret, error) {
+	secretBytes, err := os.ReadFile(filename)
+	if err != nil {
+		return nil, err
+	}
+	var secret QuaySecret
+	err = json.Unmarshal(secretBytes, &secret)
+	if err != nil {
+		return nil, err
+	}
+
+	return &secret, nil
+}
+
+func filterTagsToSync(src, target []string) []string {
+	var tagsToSync []string
+
+	targetMap := make(map[string]bool)
+	for _, targetTag := range target {
+		targetMap[targetTag] = true
+	}
+
+	for _, srcTag := range src {
+		if _, ok := targetMap[srcTag]; !ok {
+			tagsToSync = append(tagsToSync, srcTag)
+		}
+	}
+	return tagsToSync
+}
+
+// DoSync syncs the images from the source registry to the target registry
+func DoSync() error {
+	cfg := NewSyncConfig()
+	Log().Infow("Syncing images", "images", cfg.Repositories, "numberoftags", cfg.NumberOfTags)
+	ctx := context.Background()
+
+	quaySecret, err := readQuaySecret(cfg.QuaySecretFile)
+	if err != nil {
+		return fmt.Errorf("error reading secret file: %w", err)
+	}
+	qr := NewQuayRegistry(cfg, quaySecret.BearerToken)
+
+	acr := NewAzureContainerRegistry(cfg)
+	acrPullSecret, err := acr.GetPullSecret(ctx)
+	if err != nil {
+		return fmt.Errorf("error getting pull secret: %w", err)
+	}
+
+	acrAuth := types.DockerAuthConfig{Username: "00000000-0000-0000-0000-000000000000", Password: acrPullSecret.RefreshToken}
+	quayAuth := types.DockerAuthConfig{Username: quaySecret.PullUsername, Password: quaySecret.PullPassword}
+
+	for _, repoName := range cfg.Repositories {
+		var srcTags, acrTags []string
+
+		baseURL := strings.Split(repoName, "/")[0]
+		repoName = strings.Join(strings.Split(repoName, "/")[1:], "/")
+		srcAuth := types.DockerAuthConfig{}
+
+		Log().Infow("Syncing repository", "repository", repoName, "baseurl", baseURL)
+
+		if baseURL == "quay.io" {
+			srcAuth = quayAuth
+			srcTags, err = qr.GetTags(ctx, repoName)
+			if err != nil {
+				return fmt.Errorf("error getting quay tags: %w", err)
+			}
+			Log().Debugw("Got tags from quay", "tags", srcTags)
+		} else {
+			oci := NewOCIRegistry(cfg, baseURL)
+			srcTags, err = oci.GetTags(ctx, repoName)
+			if err != nil {
+				return fmt.Errorf("error getting oci tags: %w", err)
+			}
+			Log().Debugw(fmt.Sprintf("Got tags from %s", baseURL), "repo", repoName, "tags", srcTags)
+		}
```

Comment:
This is a little brittle, we are assuming here that a quay pull secret is needed for `quay.io` images and no secret is needed for any other registry right?

Reply:
See below comment, it will work using passed in containers auth file